proposals/p2360.md

```diff
@@ -0,0 +1,291 @@
+    -   This gives rise to terms such as _generic type parameter_ and
+        _associated generic type_.
+    -   It is worth noting that a generic type is not necessarily a type, as its
+        type may be an interface or constraint rather than `Type`.
+
+### Compound member access
+
+The syntax `x.(Y)`, where `Y` names an instance member, now always performs
+instance binding. Therefore, for a suitable `DebugPrintable`:
+
+-   `1.(DebugPrintable.Print)()` still prints `1`, as before.
+-   `i32.(DebugPrintable.Print)()` now prints `i32`, rather than forming a
+    member name.
+-   `1.(i32.(DebugPrintable.Print))()` is now an error, rather than printing
+    `1`.
+
+In order to get the old effect of `x.(MyType.(MyInterface.InstanceMember))()`,
+one can now write `x.((MyType as MyInterface).InstanceMember)()`. This behaves
+the same as the member access in:
+
+```
+fn F[MyType:! MyInterface, X:! Type](x: X) -> auto {
+  return x.(MyType.InstanceMember)();
+}
+```
+
+... because it is the same thing. `MyType as InstanceMember` is a facet, which
+roughly corresponds to a fully-instanstiated `impl`, and lookup into a facet
+finds members of that corresponding `impl`. And if `MyType` is declared as
+`MyType:! MyInterface` instead of as `MyType:! Type`, then of course `MyType`
+and `MyType as MyInterface` are equivalent.
+
+### Tuple and struct literals
+
+A tuple literal such as `(1, 2, 3)` produces a tuple value. Its type cannot be
+written directly as a literal: the literal `(i32, i32, i32)` also produces a
+tuple value, not a type value. However, the type of `(1, 2, 3)` is easy to
+express: it is the result of implicitly converting `(i32, i32, i32)` to a type,
+so for example `(i32, i32, i32) as Type` evaluates to the type of `(1, 2, 3)`.
+
+There is no conversion from a tuple type to a tuple value, because tuple types
+are values of type `Type`, and the type `Type` does not implicitly convert to a
+tuple type. For metaprogramming, it is likely that we will benefit from having a
+mechanism to convert a tuple type into a tuple of types, but this is likely to
+be possible using a variadic:
+
+```
+@'''
+Takes a tuple of values. Returns a tuple containing the types of those values.
+'''
+fn TupleTypeToTupleOfTypes[Types:! Type,...](x: (Types,...)) -> auto {
+  return (Types,...);
+}
+```
+
+### Name lookup into values with constrained types
+
+Given a generic function with a generic type parameter, uses of that parameter
+will be implicitly converted to type `Type`. For example:
+
+```
+fn F[T:! Printable](x: T) { x.Print(); }
+```
+
+... is equivalent to ...
+
+```
+fn F[T:! Printable](x: T as Type) { x.Print(); }
+```
+
+As a result, we can no longer describe lookup for `x` as looking into the
+type-of-type, that is, the type of the expression after `x:`, because after
+implicit conversion, that expression is of type `Type`. Instead, lookup will
+look into the type of `x`, which symbolically evaluates to `T`, and will look at
```

Comment:
This seems to be glossing over something. Didn't we just say the type of `x` is `T as Type` not `T`?

Reply:
Slightly rephrased. The point here is that we're performing a symbolic evaluation, and the result is... however the value `T` is symbolically represented as a value of type `Type`. We don't have the terminology to talk about that right now, I don't think, but I can hand-wave a bit better than I was :)